user-docs/configuring-a-site.md
# Configuring a site

This is an outline of the steps required to set up `radon-monitor` on a new computer.  These steps do not need to be followed precisely, but if we can keep the setup roughly the same at different sites it should make things easier for us later.


## Windows

### Entirely new computer

1. Install the ancillary software listed below
2. Install Radon Detector Monitor (RDM) from https://github.com/anstoradonlab/radon-monitor (look for releases to find a Windows installer) 
3. Create the directory `c:\data`
4. Choose a site name, and a name for each radon detector.  Make sure that both the site name and detector name will uniquely identify the data stream in the event that data from many detectors is merged together.  Also, choose names which can be used as file names without surprises (use no whitespace, or special characters).  If there is only one radon detector at the site, it is Ok for both the site and detector name to be the same, e.g. `GunnPoint`. A more complex example is a site with a tower.  This would be Ok for Lucas Heights with radon detectors drawing air from 2m and 50m AGL:
    * site: LucasHeights
    * detector1: LH-2m
    * detector2: LH-50m
5. Create a configuration file `c:\data\rdm-config.ini` and edit it based on the template below
    * to work out which COM port the datalogger is connected to, use <kbd>Win</kbd> + <kbd>R</kbd> to run `devmgmt.msc`.  Expand the device tree to find out which COM port is associated with any USB-Serial convertor which you might be using.
        * alternatively, the list of COM ports is available in `RDM` under `View → System Information`
6. Run RDM, from the Windows start menu
7. Ensure RDM is set as a startup program, according to [Microsoft's instructions](https://support.microsoft.com/en-us/windows/add-an-app-to-run-automatically-at-startup-in-windows-10-150da165-dcd9-7230-517b-cf3c295d89dd):
    * Select the Start button <kbd>Win</kbd> and scroll to find the app you want to run at startup.
    * Right-click the app, select More, and then select Open file location. This opens the location where the shortcut to the app is saved. If there isn't an option for Open file location, it means the app can't run at startup.
    * With the file location open, press the Windows logo key <kbd>Win</kbd> + <kbd>R</kbd>, type shell:startup, then select OK. This opens the Startup folder.
    * Copy and paste the shortcut to the app from the file location to the Startup folder.


### Conversion from a site running the VisualBasic version of RDM

1. Stop the old Radon Monitor software and remove it from the list of startup programs
2. Backup all data (the datalogger memory is going to be wiped in this procedure)
3. Follow steps 1-5 for a new install.  If you want to keep producing CSV files in the same format as the old logging system, pay attention to the `csv_file_pattern` section of the configuration file. It might also be helpful to use the short code from the old logger as the detector name, e.g. short codes like "GP", "CG", in the past. 
4. Using `Device Configuration Utility`, modify the datalogger code, as detailed below (in Appendix 2). This is optional, but will make the data more useful.
5. Also using `Device Configuration Utility`, set the datalogger clock to UTC.
6. Run RDM, from the Windows start menu
    * Go to `File → Load Configuration` and choose `c:\data\rdm-config.ini`
        * (expect logging to begin now)
    * Go to `View → Calibration` and enable scheduled calibrations
7. Ensure RDM is set as a startup program

## Linux, using `uv`

The [uv](https://docs.astral.sh/uv/getting-started/installation/) package manager is currently the recommended way to install RDM on a linux machine.  It installs Python software into an isolated environment and avoids using the system Python interpreter, so there is a decent chance that RDM will work on both old and recent distributions.

1. install uv, following [the instructions]((https://docs.astral.sh/uv/getting-started/installation/)).  TL;DR, try this:

```sh
curl -LsSf https://astral.sh/uv/install.sh | sh
```

2. install the latest version of RDM as a `uv` tool

```sh
uv tool install --python 3.11 "ansto_radon_monitor[gui] @ git+https://github.com/anstoradonlab/radon-monitor.git"
```

Once the install has finished, if it works, you can:
* check that the tool is installed, `uw tool list`
* upgrade to a new version, `uw tool upgrade`
* etc., see `uw tool --help`

To check that the installation worked completely, run

```sh
radon-monitor -h
```

to test the command line interface, and 

```sh
radon-monitor gui
```

to test the GUI.

3. To upgrade to the latest release, run

```sh
uv tool update ansto_radon_monitor
```



## Linux, using a virtual environment, including Raspberry Pi

Assuming that there is a compatible version of Python installed on the system, it should be fine to use a [virtual environment](https://docs.python.org/3/library/venv.html).  Currently, this is Python version 3.11 or greater.

If your system has an older version of Python installed, you could instead install a conda environment and then follow these steps from 2 onwards.

These instructions assume the following path names:
 - `/home/radon-logger` is the user's home directory
 - `/home/radon-logger/venv-rdm` is the Python virtual environment created during installation
 - `/home/radon-logger/data` is for the configuration file and output data
 - `/home/radon-logger/data`

0.  Install pre-requesites (prefer system packages over pypi/pip-installed packages)
```sh
sudo apt install python3-pyqt5 python3-numpy python3-msgpack python3-gevent
```

1. Initialise and then activate a new virtual environment:  
```sh
   python3 -m venv --system-site-packages ~/venv-rdm
   source ~/venv-rdm/bin/activate
```

2. Install the radon-monitor command line client and GUI.  Replace the tag (`@10.17.0`) with the version you want to install:  
```sh
python -m pip install  "ansto_radon_monitor[gui] @ git+https://github.com/anstoradonlab/radon-monitor.git@v10.17.0"
```

3. Install the Labjack exodriver (this is only necessary for the older calibration systems which contain a LabJack).  For Ubuntu or Raspberry Pi systems, execute these commands, or [read the full installation instructions](https://github.com/labjack/exodriver/blob/master/INSTALL.Linux):
```sh
sudo apt install build-essential libusb-1.0-0-dev
git clone https://github.com/labjack/exodriver.git
cd exodriver
sudo ./install.sh
```

4. You should now be able to test that the installation worked by running:  
    `~/venv-rdm/bin/radon-monitor -h`

5. At this point, the installation is complete.  You can now create a configuration file.  

6. If you are not using the GUI to schedule tasks, you can integrate the command-line controls with the system.  We can cheat by using cron jobs, for example run `crontab -e` and then set up something like this:

```bash
# Try to start logging every 10 minutes.  This will fail immediately if there is
# already a version of radon-monitor running, so it is a quick-and-dirty way to
# keep the logging process going.  It would be more appropriate to do this
# with a system service, for example a .service file.
*/10 * * * * /home/radon-logger/bin/radon-monitor -c /home/radon-logger/data/richmond-config.ini run
# Start running a calibration sequence at 5am on the 10th of each month
0 5 10 * * /home/radon-logger/bin/radon-monitor -c /home/radon-logger/data/richmond-config.ini calibrate
# Start running a background sequence at 0000 on the 20th of every third month
0 0 20 */3 * /home/radon-logger/bin/radon-monitor -c /home/radon-logger/data/richmond-config.ini background

```
7. When new versions of `radon-monitor` get released, upgrade using this command (replacing `X.xx.x` with the current version):
   `source ~/venv-rdm
   `python -m pip install --upgrade  "ansto_radon_monitor[gui] @ git+https://github.com/anstoradonlab/radon-monitor.git@vX.xx.x"`


## Troubleshooting ideas

* Some odd problems occur if the logger is not set to UTC.  If you know that the clock on the PC is set correctly, you can force the logger to sync to the PC clock using `Device Configuration Utility` or by using the `View → System Information` dialog in RDM.
* Data locations may need to exist and be writable by the RDM user
* Sometimes we have seen a crash where the COM port has stayed locked, meaning that RDM is unable to access it after it is reopened.  Try rebooting the computer.


## Appendix 1: Sample configuration file

This is a simple configuration file intended for a site with one radon detector.  The detector is called `TEST` and the datalogger is assumed to be connected to `COM3`.

```ini
# Configuration for ANSTO two filter radon detector
# Save this file to c:\data\rdm-config.ini
# Settings which will need changing are:
#   * detector name
#   * detector COM port
#   * radon source information
#   * uncomment FTP section, if you want to try using FTP backup
#      * set the FTP upload directory
#
# Basic config, one 1500 L detector
[data]
data_dir=c:\data
logfile=c:\data\radon_monitor_messages.log
# allowed log levels (from least output through to most output):
# "CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"
#loglevel="DEBUG"
loglevel= INFO
# If True, this option adds an "ApproxRadon" to the legacy format
# csv output (optional, True by default)
# legacy_file_write_approx_radon=True
#
# Network Time Protocol - check clock against this server
# Note: This is just a check, RDM will not update the computer clock
ntp_server=pool.ntp.org
#
# If the datalogger clock is set to anything other than UTC
# then set this to the correct offset (in hours).  It is 
# recommended to leave the datalogger set to UTC whenever possible
# datalogger_time_offset=0.0

# detector_kind sets the type of radon detector.  Options are:
#  "L1500" - standard 1500 litre detector
#  "L700" - standard 700 litre detector
#  "L200" - standard 200 litre detector
#  "mock" - used for testing without a connected radon detector
[detector1]
kind=L1500
serial_port=COM3
# Default baudrate is 9600.  If you have a short cable, 
# and good connect, this can be set to 115200
# baudrate=9600
#
# suggested template: 
# [SITECODE]
# or (if there is more than one detector here)
# [SITECODE]_[InletHeight or Detector Name or other identifier]
name=TEST
csv_file_pattern=height02/{NAME}{MONTH}{YEAR}.CSV
datalogger_serial=-1
# Metadata - setting this correctly is helpful for automated
# data processing later
volume_m3=1.5
thoron_delay_volume_m3=0.2
thoron_delay_number_of_tanks=2

# add more detectors like this
# [detector2]
# kind=mock
# serial_port=COM6
# name=TEST_050M
# datalogger_serial=-1
# csv_file_pattern=height50/{NAME}{MONTH}{YEAR}.CSV


[calbox]
# there are different kinds of calibration units.  These are
# - `none`: no calibration unit attached, e.g. for manual calibrations
# - `generic`: single detector, built around a Labjack connected over USB
# - `CapeGrim`: two detectors, built around a Labjack connected over USB
# - `BurkertModel1`: one or two detectors, a new system available from Jan 2023.
#                    This is interfaced over ethernet, with a Burkert ME43 gateway
# - `BurkertModel1-2b`: one or two detectors including two background control,
#                    The default for new installations from 2024 onwards
#                    This is interfaced over ethernet, with a Burkert ME43 gateway`
# There is also a kind called "mock" which can be used for testing the software
# but which doesn't try to connect to hardware.
kind=BurkertModel1-2b
# Activity of the radon source (activity of radon's parent, radium-226), 
# in decays per second (Bq)
radon_source_activity_bq=10e3

# The labjack ID can be set in software.  A special value of -1 means
# to use the first labjack found, which only makes sense if there is
# only ever one labjack connected to the PC (only applies to Labjack-based devices)
labjack_id=-1
# Only use the labjack with the serial number specifed here.  A special
# value of -1 means to ignore the serial number.
# (only applies to Labjack-based devices)
labjack_serial=-1
# Flushing duration to use during calibrations (seconds)
flush_duration_sec=3600
# Radon injection duration to use during calibrations (seconds)
inject_duration_sec=3600
# Duration to switch to background mode for background charactisation (seconds)
background_duration_sec=3600
# ip address of the Burkert ME43 gateway (only applies when kind=BurkertModel1)
me43_ip_address=192.168.0.100
# flow rate setpoint on the calibration unit's mass flow controller during source flushing (units of standard l/min)
# (only applies when kind=BurkertModel1, optional, default=0.5)
flush_flow_rate=0.5 
# flow rate setpoint on the calibration unit's mass flow controller during injection (units of standard l/min)
# (only applies when kind=BurkertModel1, optional, default=0.5)

CHANGELOG.md
# Changelog

## Version 10.20

2 July 2026

 - Increase maximum limits (to 9999 hrs) for the duration of calibration,
   background, and flush so that users can easily put the instrument into 
   one of these states for a long period of time.



## Version 10.19

2 February 2026

 - Handle calibration system with three radon detectors
 - Log the datalogger table definitions during startup
 - Special case handling for the RTV table when it contains only one
   data record, because the gemeral-case data download code can sometimes
   fail to retrieve any data (seemingly, only if the datalogger timezone is set to something other than UTC)


## Version 10.18

11 August 2025

 - Better handling for the new external flow rate sensor in plots
 - Add the option to perform calibration / backgrounds on a monthly
   schedule

## Version 10.17.2

 - Increase the amount of log messages during datalogger connection
 - Add an option to report pakbus activity

## Version 10.17.1

8 November 2024

 - Fix a problem with the sensitivity sweep dialog

## Version 10.17

16 October 2024

 - Adds the option to connect to radon detector's internal datalogger over ethernet 
   instead of the serial port.
   This is only supported for dataloggers without pakbus encryption and without
   a security code, so only use it over trusted networks.  To configure this,
   remove the `serial_port` key from the `detector` configuration and instead
   add `network_address=[ip address or hostname]`.  If the datalogger is re-configured
   to use a non-standard port, also add `network_port=[port]`.  For example, instead of: 

        [detector1]
        serial_port=COM1

    use something like this instead (using your own IP address): 

        [detector1]
        network_address=192.168.1.100

    By default, `network_port` is set to PakBus default of `6785` and does not need
    to be specified.

 - Fix GUI and command-line conflicts on Linux.  As an interim solution, now, on Linux:
    - The command line can be used to start a background process which performs
      logging without requiring a GUI, for example on a headless machine, but
    - Starting the GUI stops the background process and instead starts logging
      from inside the GUI
    - Stopping the GUI stops logging, and the user needs to restart logging 
      from the command line.

## Version 10.16

7 August 2024

 - Transparently handle data from kinds of external flow sensors and rapidly update the 
   flow rate display for the sensors which allow this
 - Remove duplicate rows when archiving data (duplicates are rare but possible)
 - Improve clock synchronisation accuracy between computer and datalogger
 - Optionally check PC clock against an independent NTP server.  Note that many
   corporate filewalls are set to block NTP traffic, so you might need to use
   your organisation's NTP server.  If enabled, datalogger clocks will only be
   updated if this check succeeds but RDM will never change the computer clock 
   To enable this option, set the NTP server  name in the configuration file, 
   for example: 

        [data]
        ntp_server=pool.ntp.org
        


## Version 10.15.1

31 July 2024

 - Change the NSIS installer so that it runs an uninstaller when installing 
   a new version into the same directory (on Windows)
 - Open serial port in exclusive mode on Linux
 - Fix a problem with command-line "query" when running on Linux

## Version 10.15

23 July 2024 

*found issues with installer so the installer was not posted to github*

 - Packaging on Windows now uses Python v3.11 and PyQt v5.15
 - Added a "scheduled tasks" view which displays RDM's internal state.  This is useful for debugging or for displaying the time until the next calibration.
 - Prevent the 'inconsistent state' message from appearing more than once per hour
 - Improve missing data handling in sensitivity sweep

## Version 10.14

18 June 2024

 - Add an experimental option to send incoming data out across the network.
   This is a feature which is intended for use with [OpenRVDAS](https://openrvdas.org/), although it could be 
   used in other situations too.  30-minute summaries from the "Results" table are sent
   as UDP packets, formatted as JSON.  An example configuration, which sends data
   to multicast group 224.0.36.0, looks like: 

        [data]
        udp_destination=224.0.36.0
        udp_port=51520
        udp_multicast_interface=<host name or ip address of the computer running RDM>

    Internally, RDM uses a vendored copy of [UDPWriter](https://github.com/OceanDataTools/openrvdas/blob/master/logger/writers/udp_writer.py) from OpenRVDAS.  It is [documented here](https://htmlpreview.github.io/?https://raw.githubusercontent.com/oceandatatools/openrvdas/master/docs/html/logger/writers/udp_writer.html). 

    If `udp_destination` is not set in the configuration file, this feature is disabled.

- Avoid crashing the GUI when the output data directory can't be created during startup

## Version 10.13

18 March 2024

- Add "gui" command line option on Linux.  (GUI is the default on Windows)
- Display pressure on main display only to one decimal place
- Add a configuration option for 700l detector: 

        [detector1]
        kind=L700

## Version 10.12

19 December 2023

-   Use thicker lines on some plots for better visibility
-   Filter FTP settings from configuration before dumping config to database
-   Bugfix: improve flag calculation so that flag no longer stays stuck in state 2

## Version 10.11

25 October 2023

-   Similar functionality to version 10.10 but the GUI code has been
    merged into the same repository as the command-line/library code.

## Version 10.10

25 October 2023

-   Sensitivty sweep: change limits on GUI inputs so that this is
    suitable for background sweep, add noise handling.
-   Bugfix: disable fitting column width for Datetime column, as this
    sometimes causes crashes.

## Version 10.9

21 September 2023


src/ansto_radon_monitor/configuration.py
"""
Configuration management.  Basic approach is similar to
https://tech.preferred.jp/en/blog/working-with-configuration-in-python/
"""

import argparse
import configparser
import copy
import datetime
import logging
from logging import StreamHandler
from logging.handlers import RotatingFileHandler
import os
import pathlib
import pprint
import sys
import typing
from dataclasses import dataclass, field
from enum import Enum

import dacite
import yaml

_logger = logging.getLogger(__name__)


from ansto_radon_monitor import __version__

LogLevel = typing.NewType("LogLevel", int)

# TODO: consider using an enumeration for detector kind?
# DetectorKind = typing.NewType("DetectorKind", str)

# DETECTOR_KIND_CHOICES = [
#     DetectorKind("L100"),
#     DetectorKind("L200"),
#     DetectorKind("L1500"),
#     DetectorKind("L5000"),
# ]


# def parse_detector_kind(s: DetectorKind) -> DetectorKind:
#     sup = DetectorKind(s.upper())
#     if not sup in DETECTOR_KIND_CHOICES:
#         raise RuntimeError(f"Unknown kind of radon detector: {s}")
#     return sup


def str2bool(v):
    if isinstance(v, bool):
        return v
    if v.lower() in ("yes", "true", "t", "y", "1"):
        return True
    elif v.lower() in ("no", "false", "f", "n", "0"):
        return False
    else:
        raise ValueError("Boolean value expected.")


@dataclass
class DetectorConfig:
    """
    Configuration of a single radon detector
    """

    name: str = ""
    serial_port: str = ""
    network_address: str = ""
    network_port: int = 6785
    baudrate: int = 9600
    kind: str = ""
    volume_m3: typing.Optional[float] = None
    thoron_delay_volume_m3: typing.Optional[float] = None
    thoron_delay_number_of_tanks: typing.Optional[float] = None
    datalogger_serial: int = -1
    csv_file_pattern: typing.Optional[str] = None
    datalogger_time_offset: float = 0.0
    report_pakbus_statistics: bool = False
    check_ntp_sync: typing.Optional[bool] = None


@dataclass
class CalUnitConfig:
    """
    Configuration of a calibration unit
    """

    kind: str = ""
    labjack_id: typing.Optional[int] = -1
    labjack_serial: int = -1
    me43_ip_address: typing.Optional[str] = None
    flush_flow_rate: float = 0.5
    inject_flow_rate: float = 0.5
    flush_duration_sec: int = 3600 * 12
    inject_duration_sec: int = 3600 * 6
    background_duration_sec: int = 3600 * 24
    radon_source_activity_bq: typing.Optional[float] = None
    flow_sensor_polynomial: typing.List[float] = field(default_factory=lambda: [0.1025, -0.17965, 0.0669979])


@dataclass
class FtpConfig:
    """
    FTP backup configuration
    """

    server: typing.Optional[str] = None
    user: typing.Optional[str] = None
    passwd: typing.Optional[str] = None
    directory: typing.Optional[str] = None


@dataclass
class Configuration:
    """
    Configuration of the entire app
    """

    foreground: bool = False
    loglevel: LogLevel = LogLevel(logging.ERROR)
    log_pakbus_activity: bool = False
    logfile: typing.Optional[pathlib.Path] = None
    pid_file: pathlib.Path = pathlib.Path("/tmp/ansto_radon_monitor.pid")
    data_dir: pathlib.Path = pathlib.Path(".", "data").absolute()
    data_file: typing.Optional[pathlib.Path] = None
    legacy_file_timezone: float = 0
    legacy_file_write_approx_radon: bool = True
    backup_time_of_day: datetime.time = datetime.time(0,10)
    detectors: typing.List[DetectorConfig] = field(default_factory=list)
    calbox: CalUnitConfig = field(default_factory=CalUnitConfig)
    ftp: FtpConfig = field(default_factory=FtpConfig)
    udp_destination: typing.Optional[str] = None
    udp_port: int = 51520
    udp_multicast_interface: typing.Optional[str] = None
    ntp_server: typing.Optional[str] = None

    
    def as_text(self, include_sensitive=False) -> str:
        """Generate a text version of the configuration

        Args:
            include_sensitive (bool, optional): 
            If *true* then include potentially senstive information. Defaults to False.

        Returns:
            str: pretty-printed text version of the configuration
        """
        config = self
        if not include_sensitive:
            config = self.as_redacted_config()
        else:
            config = self
        return pprint.pformat(config)
    
    def as_redacted_config(self):
        """
        Return a copy of the configuration with sensitive information removed
        """
        config = copy.deepcopy(self)
        if config.ftp.server is not None:
            config.ftp.server = "XXXXXXXX"
        if config.ftp.user is not None:
            config.ftp.user = "XXXXXXXX"
        if config.ftp.passwd is not None:
            config.ftp.passwd = "XXXXXXXX"
        return config
        

def parse_config(raw_cfg) -> Configuration:
    # define converters/validators for the various data types we use
    # a dict mapping a type to a convertor function
    converters = {
        pathlib.Path: lambda x: pathlib.Path(x).absolute(),
        int: int,
        LogLevel: lambda x: LogLevel(logging._nameToLevel[x]),
        #        DetectorKind: parse_detector_kind,
        bool: str2bool,
        float: float,
        datetime.time: lambda x: datetime.datetime.strptime(x, '%H:%M').time(),
        typing.List[float]: lambda x: [float(itm) for itm in x.strip('[]').split(',')],
    }

    # create and validate the Configuration object
    configuration = dacite.from_dict(
        data_class=Configuration,
        data=raw_cfg,
        config=dacite.Config(type_hooks=converters),
    )

    # handle data_file not set (which is probably the usual case)
    if configuration.data_file is None:
        configuration.data_file = pathlib.Path(
            configuration.data_dir, "current", "radon.db"
        )
    
    # For the ntp server configuration, set the per-detector
    # flags if the user has set a ntp server address
    
    for ii, detconfig in enumerate(configuration.detectors):
        if detconfig.check_ntp_sync is None:
            detconfig.check_ntp_sync = ((configuration.ntp_server is not None) and 
                                        (configuration.ntp_server != ""))
        # Validation: do not accept both a serial port and a network address for the same
        # detector
        if detconfig.serial_port != "" and detconfig.network_address != "":
            error_message = f"Specifiy either 'serial_port' OR 'network_address' but not both; detector {ii+1} has been configured with 'serial_port = {detconfig.serial_port}' and 'network_address = {detconfig.network_address}'"
            _logger.error(error_message)
            raise RuntimeError(error_message)
        
    # Validation - check that no two detectors have the same serial port
    seen_ports = []
    seen_ip_addresses = []
    for ii, detconfig in enumerate(configuration.detectors):
        if detconfig.serial_port != "":
            if detconfig.serial_port in seen_ports:
                error_message = f"Make sure all radon detectors are configured with uniquie serial ports; Detector {ii+1} has 'serial_port {detconfig.serial_port}' but this has already been used for another detector."
                _logger.error(error_message)
                raise RuntimeError(error_message)
            else:
                seen_ports.append(detconfig.serial_port)

        if detconfig.network_address != "":
            if detconfig.network_address in seen_ip_addresses:
                error_message = f"Make sure all radon detectors are configured with uniquie network addresses; Detector {ii+1} has 'network_address {detconfig.network_address}' but this has already been used for another detector."
                _logger.error(error_message)
                raise RuntimeError(error_message)
            else:
                seen_ip_addresses.append(detconfig.network_address)

    return configuration


def get_parser():
    parser = argparse.ArgumentParser(
        description="Control and monitoring for ANSTO radon detectors",
        formatter_class=argparse.RawTextHelpFormatter
    )

    parser.add_argument(
        "-c",
        "--config",
        type=lambda p: pathlib.Path(p).absolute(),
        default=pathlib.Path(pathlib.Path(".").absolute(), "config.yaml"),
        help="Name of configuration file",
        dest="configuration_file",
    )

    parser.add_argument(
        "--version",
        action="version",
        version="ansto_radon_monitor {ver}".format(ver=__version__),
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const="INFO",
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const="DEBUG",
    )

    parser.add_argument(
        "-fg",
        "--foreground",
        dest="foreground",
        help="Run in foreground",
        default="False",
        action="store_true",
    )

    parser.add_argument(
        "action",
        choices=[
            "run",
            "gui",
            "query",
            "quit",
            "calibrate",
            "background",
            "listserialports",
            "listlabjacks",
        ],
        default="run",
        help="""The action to perform
    
COMMAND-LINE RADON DETECTOR ACTIONS
These actions need a configuration file to be specified (using the -c or 
--config option)

    run
    Run the monitoring process in the background (or, if used with the -fg or
    --foreground option, run in the foreground)
   
    query
    Display status of the monitoring process and exit
    
    quit
    Terminate the monitoring process
    
    calibrate
    Ask the monitoring process to begin a calibration sequence
    
    background
    Ask the monitoring process to  begin a background sequence

GUI OPTION
This action does not require a configuration file to be specified

    gui
    Start the GUI.  The GUI will re-open the configuration file which
    was most recently used from the GUI.  If the GUI was logging
    when it was last used, then logging will resume.

AUXILLARY ACTIONS
These actions do not require a configuration file

    listserialports
    List all serial ports on the system

    listlabjacks
    List all labjack U12s connected to the system
    """,
    )
    return parser


def parse_args(args: typing.List[str]):
    """Parse command line parameters

    Args:
        args ([str]): command line parameters as list of strings

    Returns:
        :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = get_parser()
    return parser.parse_args(args)


def config_from_commandline(
    args: typing.List[str], raw_cfg: typing.Union[dict, None] = None
):
    """Load the application configuration, based on command line options

    Parameters
    ----------
    args : list[str]
        command line arguments

    raw_cfg : dict, optional
        raw configuration, optional.  If present, then `raw_cfg` is used instead of reading from
        a configuration file.
    """

    cmdline_args = parse_args(args)
    print(args, cmdline_args)

    if raw_cfg is None:
        if cmdline_args.configuration_file.exists():
            raw_cfg = raw_config_from_inifile(cmdline_args.configuration_file)
        else:
            _logger.error(
                f'Configuration file "{cmdline_args.configuration_file}" does not exist.'
            )
            get_parser().print_help(sys.stderr)
            sys.exit(1)
    else:
        raw_cfg = copy.deepcopy(raw_cfg)

    # over-write config where options have been specified on the command line
    for k in ["loglevel", "foreground"]:
        val = vars(cmdline_args)[k]
        if val is not None:
            raw_cfg[k] = val

    print(raw_cfg)

    config = parse_config(raw_cfg)

    # validate configuration
    # (doing this early seems helpful, but may not be ideal in some respects.
    # Consider moving it later, e.g. when DataStore is initialized)
    if not config.data_dir.exists():
        _logger.error(f'Data storage directory "{config.data_dir}" does not exist.')

    # check log file is writable
    if config.logfile is None:
        config.logfile = config.data_dir.joinpath("radon_monitor_messages.log")
    try:
        with open(config.logfile, "at") as fd:
            pass
    except:
        _logger.error(
            f"Unable to open log file for writing, log messages will"
            f" not be saved.  Filename: {config.logfile}"
        )
        config.logfile = None
    _logger.debug(f"Configuration parsed: {config}")

    return config, cmdline_args


def config_from_yamlfile(filename) -> Configuration:
    filename = os.path.normpath(filename)
    _logger.info(f"Loading configuration from: {filename}")
    with open(filename, "rt") as fd:
        raw_cfg = yaml.safe_load(fd.read())
    config = parse_config(raw_cfg)
    return config


def config_from_inifile(filename) -> Configuration:
    """Read configuration from an ini file

    Parameters
    ----------
    filename : path
        filename, e.g. config.ini

    Returns
    -------
    Configuration
        Parsed configuration
    
    Note
    ----

    The ini file is assumed to have sections [data], [calbox], and
    one or more [detector...] sections (e.g. [detector1], [detector2])

    """
    raw_cfg = raw_config_from_inifile(filename)
    config = parse_config(raw_cfg)
    return config

def raw_config_from_inifile(filename) -> typing.Dict:
    """read configuration from inifile as a dict
    See also config_from_inifile

    Parameters
    ----------
    filename : str
        file path

    Returns
    -------
    Dict
        dict of configuration
    """
    filename = os.path.normpath(filename)
    _logger.info(f"Loading configuration from: {filename}")
    configp = configparser.ConfigParser()
    configp.read(filename)
    # load the detectors into an array of dicts
    detectors_config = []
    raw_cfg: typing.Dict[str, typing.Any] = {}
    for section in configp.sections():
        if section.startswith("detector"):
            detectors_config.append(dict(configp[section]))
        # config items under "data" are translated to top-level keys
        elif section == "data":
            raw_cfg.update(configp["data"])
        else:
            raw_cfg[section] = dict(configp[section])
    raw_cfg["detectors"] = detectors_config
    return raw_cfg


if __name__ == "__main__":
    raw_cfg = {
        "detector_config": [
            {"kind": "L1500", "name": "low", "port": "/dev/ttyS0"},
            {"kind": "L1500", "name": "high", "port": "/dev/ttyS1"},
        ],
        "number_of_detectors": "2",
        "loglevel": "ERROR",
    }
    configuration = parse_config(raw_cfg)

    import pprint

    pprint.pprint(configuration)

    args = ["-vv", "--config=test.yaml"]
    print(parse_args(args))

    args = ["-vv", "--config=__USE_DUMMY_INTERNAL_CONFIG"]
    config = config_from_commandline(args)
    assert config.loglevel == logging.DEBUG

    args = ["--config=__USE_DUMMY_INTERNAL_CONFIG"]
    config = config_from_commandline(args)
    assert config.loglevel == logging.ERROR


def setup_logging(loglevel=logging.DEBUG, logfn=None, log_pakbus_activity=False):
    """Setup basic logging

    Args:
        loglevel (int): minimum loglevel for emitting messages
        logfn (str): file name to log messages to (if None, don't log messages to file)

    """

    if log_pakbus_activity:
        pakbus_loglevel = loglevel
    else:
        pakbus_loglevel = logging.CRITICAL

    # exclude information messages for Pylink and Pycr1000
    for mod in ["pycampbellcr1000", "pylink"]:
        logging.getLogger(mod).setLevel(pakbus_loglevel)

    rootlogger = logging.getLogger()
    rootlogger.setLevel(loglevel)
    # logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    log_format = "[%(levelname)1.1s %(asctime)s %(module)s:%(lineno)d %(threadName)s] %(message)s"
    datefmt = "%Y-%m-%d %H:%M:%S%z"

    # add a stream handler if non is present
    if len(rootlogger.handlers) == 0:
        stream_handler = StreamHandler(sys.stderr)
        stream_handler.setLevel(loglevel)
        rootlogger.addHandler(stream_handler)

    if logfn is not None:
        # check for existing RotatingFileHandlers and remove them
        for hdlr in rootlogger.handlers:
            if isinstance(hdlr, RotatingFileHandler):
                rootlogger.removeHandler(hdlr)

        file_handler = RotatingFileHandler(
            logfn,
            mode="a",
            maxBytes=5 * 1024 * 1024,
            backupCount=10,
            encoding=None,
            delay=0,
        )
        log_formatter = logging.Formatter(log_format, datefmt=datefmt)
        file_handler.setFormatter(log_formatter)
        file_handler.setLevel(loglevel)
        rootlogger.addHandler(file_handler)

    log_formatter = logging.Formatter(log_format, datefmt=datefmt)
    for hdlr in rootlogger.handlers:
        hdlr.formatter = log_formatter

